server/routerlicious/packages/services-telemetry/src/lumberjack.ts

```diff
@@ -0,0 +1,111 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import path from "path";
+import { LumberEventName } from "./lumberEventNames";
+import { Lumber } from "./lumber";
+import { LogLevel, LumberType, ITelemetryMetadata, ILumberjackEngine } from "./resources";
+
+// Lumberjack is a telemetry manager class that allows the collection of metrics and logs
+// throughout the service. A list of ILumberjackEngine must be provided to Lumberjack
+// by calling setupEngines() before Lumberjack can be used.
+export class Lumberjack {
+    protected static _instance: Lumberjack | undefined = undefined;
+    private readonly engineList: ILumberjackEngine[];
+    private isSetupCompleted: boolean;
+
+    protected constructor() {
+        this.engineList = [];
+        this.isSetupCompleted = false;
+    }
+
+    public static get instance() {
+        if (!Lumberjack._instance) {
+            Lumberjack._instance = new Lumberjack();
+        }
+
+        return Lumberjack._instance;
+    }
+
+    public setupEngines(engines: ILumberjackEngine[]) {
+        if (this.isSetupCompleted) {
+            throw new Error("This Lumberjack was already setup with a list of engines.");
+        }
+
+        this.engineList.push(...engines);
+        this.isSetupCompleted = true;
+    }
+
+    public newLumberMetric<T extends string = LumberEventName>(eventName: T) {
+        this.throwOnEmptyEngineList();
+        return new Lumber<T>(eventName, LumberType.Metric, this.engineList);
+    }
+
+    public log(
+        message: string,
+        metadata: ITelemetryMetadata,
+        level: LogLevel,
+        properties?: Map<string, any> | Record<string, any> | undefined,
```

Comment:
I decided to store properties in `Lumber` using `Map` because it has [some advantages over `Object`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps): for example, accidental keys (while `Object`s have a prototype), iteration and performance, such as frequent addition and removals which could happen by chaining multiple `addProperty()` calls in `Lumber`. Then, I thought: if I am managing properties using a `Map`, maybe the user will be doing the same. So it would be great to support a properties parameter that can either be a `Map` or a `Record`/`Object`.

But I agree with you that the implementation was missing a constructor parameter for an existing properties array, or an addProperties() calls that takes a whole set of properties at once. I am adding that now.